Notify when Manager.from_queryset happens inside model class body

flaeppe opened the pull request:
Whenever `Manager.from_queryset` happens inside of a model class body, an error will be displayed

e.g. Doing

```python
class MyModel(models.Model):
    objects = BaseManager.from_queryset(MyQuerySet)()
```

Yields:

```
error: `.from_queryset` called from inside model class body
```

---
__Note:__ I've also included a refactoring commit for config parsing, that hopefully will make it less of a hassle to introduce new settings in the future.

## Related issues

Refs #738

flaeppe committed:
Refactor to more easily support additional config options

flaeppe committed:
Notify when Manager.from_queryset happens inside model class body

- A warning will be emitted whenever `Manager.from_queryset` happens
  inside of a model class body

sobolevn commented on README.md:
```diff
@@ -66,6 +66,16 @@ We rely on different `django` and `mypy` versions:
+
+An optional strictness flag is available for stronger checks:
+
+- `warn_manager_type_created_in_model_body`
```
We don't need to make this a setting. Since this error can be fixed with a simple `# type: ignore` comment.

flaeppe replied:
Alright, I'll drop the setting

sobolevn commented on mypy_django_plugin/config.py:
```diff
@@ -0,0 +1,111 @@
+        else:
+            self.parse_ini_file(filepath)
+
+    def parse_toml_file(self, filepath: Path) -> None:
```
I am +1 on using a separate module for this, but without the new setting 🙂

sobolevn reviewed (changes requested):
Awesome, thank you!

flaeppe committed:
Resolve generated default manager types before final iteration

A default manager on a model should always exist, eventually. Although,
we extend to look through dynamically generated managers on each
iteration instead of deferring until the final iteration.

flaeppe commented on tests/typecheck/fields/test_related.yml:
```diff
@@ -608,8 +608,12 @@
+        reveal_type(user.book_set.add)  # N: Revealed type is "def (*objs: Union[myapp.models.Book*, builtins.int], *, bulk: builtins.bool =)"
+        reveal_type(user.article_set.add)  # N: Revealed type is "def (*objs: Union[myapp.models.Article*, builtins.int], *, bulk: builtins.bool =)"
+        reveal_type(user.book_set.queryset_method())  # N: Revealed type is "builtins.int"
+        reveal_type(user.article_set.queryset_method())  # N: Revealed type is "builtins.int"
```
I've extended this test to also check in on methods only existing on the custom queryset class. Basically to ensure the related manager is built from correct bases. I realised the bases were previously incorrect, so the related manager was built from `models.Manager` and `RelatedManager` which yielded an `attribute does not exist` error for any methods existing on the custom queryset.

I'm thinking that the asserts now should ensure a correct hierarchy. Perhaps we should also check on e.g. `user.book_set.filter` to see that the `Manager`/`BaseManager` is also a base and correct? Since now we're checking that methods from `RelatedManager` and `LibraryEntityQuerySet` are included

sobolevn replied:
> I'm thinking that the asserts now should ensure a correct hierarchy. Perhaps we should also check on e.g. user.book_set.filter to see that the Manager/BaseManager is also a base and correct?

Sure, why not?

flaeppe commented on mypy_django_plugin/transformers/models.py:
```diff
@@ -326,10 +326,8 @@ def run_with_model_cls(self, model_cls: Type[Model]) -> None:
-                    # See: https://docs.djangoproject.com/en/dev/topics/db/queries/#using-a-custom-reverse-manager
-                    objects = related_model_info.get("objects")
-                    if not objects:
+                    default_manager = related_model_info.get("_default_manager")
+                    if not default_manager:
```
Changed from `objects` to `_default_manager` with https://docs.djangoproject.com/en/dev/topics/db/queries/#using-a-custom-reverse-manager as justification ("By default the RelatedManager used for reverse relations is a subclass of the default manager for that model.")

flaeppe commented on mypy_django_plugin/transformers/models.py:
```diff
@@ -360,7 +358,7 @@ def run_with_model_cls(self, model_cls: Type[Model]) -> None:
         self, related_model_cls: Type[Model], related_model_info: TypeInfo
     ) -> Optional[Instance]:
-        manager = related_model_cls._meta.managers_map["objects"]
+        manager = related_model_cls._meta.managers_map["_default_manager"]
```
Changed from `objects` to `_default_manager` with https://docs.djangoproject.com/en/dev/topics/db/queries/#using-a-custom-reverse-manager as justification ("By default the RelatedManager used for reverse relations is a subclass of the default manager for that model.")

sobolevn reviewed (approved):
Thank you!
